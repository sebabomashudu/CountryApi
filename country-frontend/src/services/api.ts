import axios from 'axios'

export interface Country {
  name: string;
  flagUrl: string;
}
/*const REST_COUNTRIES_API = 'https://restcountries.com/v3.1'

export const fetchAllCountries = async () => {
  const response = await axios.get(`${REST_COUNTRIES_API}/all`)
  return response.data
}

export const fetchCountryByName = async (name: string) => {
  const response = await axios.get(`${REST_COUNTRIES_API}/name/${name}`)
  return response.data[0] // Return first match
}*/


export const fetchAllCountries = async () => {
  const response = await axios.get('/api/Countries') // Will be proxied to C# API
  return response.data
}
export const fetchCountryByName = async (name: string) => {
  const response = await axios.get(`/api/Countries/${name}`)
  return response.data // first match
}
