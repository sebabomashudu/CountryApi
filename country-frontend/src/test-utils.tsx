// src/test-utils.tsx
import { MemoryRouter } from 'react-router-dom';
import type { ReactNode } from 'react';  // Note the 'type' keyword here

export function TestWrapper({ children }: { children: ReactNode }) {
  return <MemoryRouter>{children}</MemoryRouter>;
}