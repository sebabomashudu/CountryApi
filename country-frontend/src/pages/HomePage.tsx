import { useEffect, useState } from 'react'
import { useNavigate } from 'react-router-dom'
import FlagCard from '../components/FlagCard'
import { fetchAllCountries } from '../services/api'


interface Country {
  name: string;
  flagUrl: string;
  population: number;
  capital: string;
}

export default function HomePage() {
  const [countries, setCountries] = useState<Country[]>([])
  const navigate = useNavigate()

  useEffect(() => {
    fetchAllCountries().then(setCountries)
  }, [])

  return (
    <div className="p-4">
      <h1 className="text-2xl font-bold mb-4">Country Flags</h1>
      <div className="grid grid-cols-2 md:grid-cols-4 gap-4">
        {countries.map((country) => (
          <FlagCard
            key={country.name}  // Using name directly as string
            country={country}
            onClick={() => navigate(`/country/${country.name}`)}
          />
        ))}
      </div>
    </div>
  )
}