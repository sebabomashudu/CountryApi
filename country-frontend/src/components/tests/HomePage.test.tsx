import { render, screen, waitFor } from '@testing-library/react';
import HomePage from '../../pages/HomePage';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { fetchAllCountries } from '../../services/api';
import { TestWrapper } from '../../test-utils';

// Mock the API module
vi.mock('../../services/api', () => ({
  fetchAllCountries: vi.fn(),
}));

describe('HomePage', () => {
  beforeEach(() => {
    // Clear all mocks between tests
    vi.clearAllMocks();
  });

  it('displays loading state initially', () => {
    // Mock an unresolved promise to test loading state
    vi.mocked(fetchAllCountries).mockReturnValue(new Promise(() => {}));
    
    render(
      <TestWrapper>
        <HomePage />
      </TestWrapper>
    );
    
    expect(screen.getByText('Loading countries...')).toBeInTheDocument();
  });

  it('displays countries after loading', async () => {
    const mockCountries = [
      { name: 'Country 1', flagUrl: 'flag1.svg' },
      { name: 'Country 2', flagUrl: 'flag2.svg' },
    ];
    
    vi.mocked(fetchAllCountries).mockResolvedValue(mockCountries);
    
    render(
      <TestWrapper>
        <HomePage />
      </TestWrapper>
    );
    
    await waitFor(() => {
      expect(screen.getByText('Country 1')).toBeInTheDocument();
      expect(screen.getByText('Country 2')).toBeInTheDocument();
    });
    
    expect(screen.queryByText('Loading countries...')).not.toBeInTheDocument();
  });

  it('displays error message when fetch fails', async () => {
    vi.mocked(fetchAllCountries).mockRejectedValue(new Error('API Error'));
    
    render(
      <TestWrapper>
        <HomePage />
      </TestWrapper>
    );
    
    await waitFor(() => {
      expect(screen.getByText('Failed to load countries')).toBeInTheDocument();
    });
    
    expect(screen.queryByText('Loading countries...')).not.toBeInTheDocument();
  });
});