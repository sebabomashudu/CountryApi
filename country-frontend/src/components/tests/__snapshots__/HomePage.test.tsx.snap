// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage > matches snapshot when loaded 1`] = `
<div>
  <div
    class="p-4"
  >
    <h1
      class="text-3xl font-bold mb-6 text-center"
    >
      Countries
    </h1>
    <div
      class="w-full"
    >
      <table
        class="w-full table-fixed"
      >
        <tbody>
          <tr>
            <td
              class="align-top p-2 border"
            >
              <a
                class="block text-center hover:bg-gray-50 p-2"
                data-discover="true"
                href="/country/Country 1"
              >
                <div
                  class="font-medium mb-2"
                >
                  Country 1
                </div>
                <img
                  alt="Country 1 flag"
                  class="mx-auto w-16 h-auto border"
                  src="flag1.svg"
                />
              </a>
            </td>
            <td
              class="align-top p-2 border"
            >
              <a
                class="block text-center hover:bg-gray-50 p-2"
                data-discover="true"
                href="/country/Country 2"
              >
                <div
                  class="font-medium mb-2"
                >
                  Country 2
                </div>
                <img
                  alt="Country 2 flag"
                  class="mx-auto w-16 h-auto border"
                  src="flag2.svg"
                />
              </a>
            </td>
            <td
              class="align-top p-2 border"
            />
            <td
              class="align-top p-2 border"
            />
            <td
              class="align-top p-2 border"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
